### ISSUE TITLE
Lumo mode

### ISSUE THREAD

**Issue by arichiardi** [clojure-emacs/inf-clojure#43]
Hello @bbatsov,

I have a brand new minor mode that I would like to put somewhere :smile: 

It is a customization of a few vars of `inf-clojure` and I need some to see what is the right approach for sharing it. I could of course putting it under my name, but it won't be that visible. So I was wondering if `clojure-emacs` is a good idea for it or you have a better one :bulb: 

Given that there is no way to distinguish automatically cljs files for self hosted from normal ones, I guess this mode could be manually enabled or we can expose some `defcustom`.

Let me know!

**Comment by bbatsov** [comment]
> It is a customization of a few vars of inf-clojure and I need some to see what is the right approach for sharing it. I could of course putting it under my name, but it won't be that visible. So I was wondering if clojure-emacs is a good idea for it or you have a better one 💡

I'm fine with hosting more packages under `clojure-emacs`, but I'm not quite certain what your mode does. It might be some functionality that we could add to `inf-clojure`.

> Given that there is no way to distinguish automatically cljs files for self hosted from normal ones, I guess this mode could be manually enabled or we can expose some defcustom.

No idea what you mean here. :-) My knowledge of cljs is pretty basic.

**Comment by arichiardi** [comment]
@bbatsov my mode just customizes `inf-clojure` stuff, it is very simple and basically only 3 main functions:

```
(defun lumo--empty-string-p (string)
  "Return true if the string is empty or nil. Expects string."
  (or (null string)
      (zerop (length string))))

(defun lumo--read-classpath (classpath-option)
  "Tries to read the classpath from the input classpath option."
  (let ((classpath-file-path (concat (inf-clojure-project-root) classpath-option)))
    (cond
     ((and (file-exists-p classpath-file-path) (file-readable-p classpath-file-path))
      (f-read classpath-file-path))
     ;; TODO launch a command that returns the classpath string?
     ((user-error (concat "Option \"" classpath-option "\" was not a (readable) file, the classpath will be empty."))
      ""))))

(defun lumo--inf-clojure-program ()
  (if (lumo--empty-string-p lumo-port)
      (let ((lumo-cmd (concat lumo-command
                              " "
                              (when (not (lumo--empty-string-p lumo-args))
                                (concat lumo-args " "))
                              "-c " (lumo--read-classpath lumo-classpath-generator)))))
    (cons lumo-hostname (if (stringp lumo-port)
                            (string-to-number lumo-port)
                          lumo-port))))
```

About the last point, given that not every `.cljs` can work for `lumo` (because it is bootstrap ClojureScript), you have to enable things manually, or through `.dir-locals.el`:

```
((nil . ((lumo-port . "5044")
         (eval . (add-hook 'clojure-mode-hook #'inf-clojure-minor-mode))
         (eval . (add-hook 'clojure-mode-hook #'lumo-minor-mode)))))
```

**Comment by bbatsov** [comment]
Hmm, as `lumo` is just a cljs REPL I don't quite understand the need for a minor mode. I'd simply add the missing functions and a few extra variables to `inf-clojure` - seems to me it'd be easier this way.

**Comment by arichiardi** [comment]
Yeah that can be good as well, then how to switch to it? Were you thinking about adding a defcustom the when set to `'lumo` sets the other defcustoms correctly?

**Comment by arichiardi** [comment]
Basically all the inf-clojure command need to point to `lumo.repl` for getting informatition (for instance the completion function is `lumo.repl/get-completions`

**Comment by bbatsov** [comment]
> Yeah that can be good as well, then how to switch to it? Were you thinking about adding a defcustom the when set to 'lumo sets the other defcustoms correctly?

I'd simply add some code to check the capabilities of the REPL (e.g. see if some ns or var is present or something), use this to set the type of the REPL (e.g. clj, cljs, lumo) and invoke the proper code based on the type of the REPL. That's much cleaner than having to fiddle with some extra minor mode IMO.

**Comment by arichiardi** [comment]
OK then, sold, will do that.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by arichiardi** [comment]
@bbatsov this approach won't work because we set things *for* launching the repl, so we effectively have not way to distinguish the `mode` we want to run things on. The `defcustom` approach would work though and I am now implementing that.

**Comment by bbatsov** [comment]
> @bbatsov this approach won't work because we set things for launching the repl, so we effectively have not way to distinguish the mode we want to run things on. The defcustom approach would work though and I am now implementing that.

For the launching part - you can either have different commands to start different types of REPLs or try to determine the type of REPL to start based on some combination of configuration and inspection of the environment. E.g. if you're in a cljs file and lumo is available you start a lumo REPL with inf-clojure

**Comment by arichiardi** [comment]
Got it, I will post a solution with an explicit `defcustom` and we'll discuss on that, I am running out of open source time now :smile:

**Comment by bbatsov** [comment]
OK.

**Comment by arichiardi** [comment]
Closed with #44
